lib/fizzy/execute.cpp

```diff
@@ -487,27 +512,55 @@ execution_result execute(Instance& instance, FuncIdx func_idx, std::vector<uint6
                     instance.module.funcsec[called_func_idx - instance.imported_functions.size()];
             assert(type_idx < instance.module.typesec.size());
 
-            const auto num_inputs = instance.module.typesec[type_idx].inputs.size();
+            if (invoke_function(type_idx, called_func_idx, instance, stack))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::call_indirect:
+        {
+            const auto expected_type_idx = read<uint32_t>(immediates);
+            assert(expected_type_idx < instance.module.typesec.size());
+
+            const auto elem_idx = stack.pop();
+            if (elem_idx >= instance.table.size())
+            {
+                trap = true;
+                goto end;
+            }
+
+            const auto called_func_idx = instance.table[elem_idx];
+            assert(called_func_idx <
+                   instance.imported_functions.size() + instance.module.funcsec.size());
+
+            // check actual type against expected type
+            const auto actual_type_idx =
+                called_func_idx < instance.imported_functions.size() ?
+                    instance.imported_function_types[called_func_idx] :
+                    instance.module.funcsec[called_func_idx - instance.imported_functions.size()];
+            assert(actual_type_idx < instance.module.typesec.size());
+            const auto& expected_type = instance.module.typesec[expected_type_idx];
+            const auto& actual_type = instance.module.typesec[actual_type_idx];
+            if (expected_type.inputs != actual_type.inputs ||
+                expected_type.outputs != actual_type.outputs)
+            {
+                trap = true;
+                goto end;
+            }
+
+            const auto num_inputs = instance.module.typesec[actual_type_idx].inputs.size();
             assert(stack.size() >= num_inputs);
             std::vector<uint64_t> call_args(
```

Comment:
Why is this done here but not transmitted to `invoke_function`?

Reply:
Oh this shouldn't be here!
some tests are missing...

Reply:
Added a test that would catch this